### PR TITLE
RETIRED: Hibernate Migration (replaced by 543)

### PR DESCRIPTION
**Pull request by RobGeada**
Many thanks for submitting your Pull Request :heart:!

Please make sure that your PR meets the following requirements:

- [ ] You have read the [contributors guide](https://github.com/trustyai-explainability/trustyai-explainability/blob/main/CONTRIBUTING.md)
- [ ] Pull Request title is properly formatted: `FAI-XYZ Subject`
- [ ] Pull Request title contains the target branch if not targeting main: `[0.3.x] FAI-XYZ Subject`
- [ ] Pull Request contains link to the JIRA issue
- [ ] Pull Request contains link to any dependent or related Pull Request
- [ ] Pull Request contains description of the issue
- [ ] Pull Request does not include fixes for issues other than the main ticket

### CODE REVIEW

**Comment by RobGeada** [comment]
/test trustyai-explainability-e2e
--


<br class="Apple-interchange-newline">

File: explainability-core/pom.xml [hunk 1]
```diff
@@ -84,8 +84,32 @@
       <groupId>org.apache.opennlp</groupId>
       <artifactId>opennlp-tools</artifactId>
     </dependency>
+    <dependency>
+      <groupId>io.quarkus</groupId>
+      <artifactId>quarkus-hibernate-orm</artifactId>
+      <version>3.2.11.Final</version>
+    </dependency>
```
Comment (danielezonca) [inline explainability-core/pom.xml:91]: I would try to keep `explainability-core` thin so I prefer to have a `explainability-persistance` module to compose this use case if possible.

File: explainability-core/pom.xml [hunk 2]
```diff
@@ -84,8 +84,32 @@
       <groupId>org.apache.opennlp</groupId>
       <artifactId>opennlp-tools</artifactId>
     </dependency>
+    <dependency>
+      <groupId>io.quarkus</groupId>
+      <artifactId>quarkus-hibernate-orm</artifactId>
+      <version>3.2.11.Final</version>
```
Comment (danielezonca) [inline explainability-core/pom.xml:90]: This should be managed via the import of quarkus bom instead of explicit version (this version is tied to quarkus version)

File: explainability-core/src/main/java/org/kie/trustyai/explainability/model/Value.java [hunk 3]
```diff
@@ -28,44 +28,54 @@
...
+import jakarta.persistence.Access;
+import jakarta.persistence.AccessType;
+import jakarta.persistence.Embeddable;
+import jakarta.persistence.Transient;
+
 /**
  * Wrapper class for any kind of value part of a prediction input or output.
  *
  */
+@Embeddable
```
Comment (danielezonca) [inline explainability-core/src/main/java/org/kie/trustyai/explainability/model/Value.java:40]: If you only need similar annotation in core you can add jakarta-persistance-api dependency

File: explainability-core/src/main/java/org/kie/trustyai/explainability/model/dataframe/DataframeColumn.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
+package org.kie.trustyai.explainability.model.dataframe;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Stream;
+
+import org.kie.trustyai.explainability.model.Value;
+
+// hibernate wrapper for an arraylist
+//@Entity
```
Comment (danielezonca) [inline explainability-core/src/main/java/org/kie/trustyai/explainability/model/dataframe/DataframeColumn.java:10]: Leftover?

File: explainability-core/src/main/java/org/kie/trustyai/explainability/model/dataframe/DataframeInternalData.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
+package org.kie.trustyai.explainability.model.dataframe;
+
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+//@Embeddable
```
Comment (danielezonca) [inline explainability-core/src/main/java/org/kie/trustyai/explainability/model/dataframe/DataframeInternalData.java:8]: Same here

File: explainability-core/src/main/java/org/kie/trustyai/explainability/model/dataframe/DataframeMetadata.java [hunk 6]
```diff
@@ -98,7 +123,7 @@ public synchronized void add(Output output) {
     private void updateCachedColumnNames() {
         List<String> newNames = new ArrayList<>();
         for (int i = 0; i < this.names.size(); i++) {
-            if (this.nameAliases.get(i) != null) {
+            if (!this.nameAliases.get(i).equals("")) {
```
Comment (danielezonca) [inline explainability-core/src/main/java/org/kie/trustyai/explainability/model/dataframe/DataframeMetadata.java:126]: To avoid potential NPE
```suggestion
            if (!"".equals(this.nameAliases.get(i)) {
```

File: explainability-service/src/test/java/org/kie/trustyai/service/data/storage/performance/PerformanceTest.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+import org.kie.trustyai.explainability.model.dataframe.Dataframe;
+import org.kie.trustyai.service.data.storage.Storage;
+import org.kie.trustyai.service.utils.DataframeGenerators;
+
+import jakarta.enterprise.inject.Instance;
+import jakarta.persistence.EntityManager;
+import jakarta.persistence.PersistenceContext;
+
+public class PerformanceTest<T extends Storage> {
```
Comment (danielezonca) [inline explainability-service/src/test/java/org/kie/trustyai/service/data/storage/performance/PerformanceTest.java:15]: We should migrate this to proper JHM test.
I would probably split this work in a different PR

File: explainability-service/src/test/java/org/kie/trustyai/service/data/storage/performance/PerformanceTest.java [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+        for (int i = 0; i < N_WARMUP; i++) {
+            testFunc.accept(i);
+        }
+        long start = System.currentTimeMillis();
+        System.out.println("warmup finished after " + (start - warmStart) / 1000. + "s");
+
+        for (int i = 0; i < N_TESTS; i++) {
+            testFunc.accept(i);
+        }
+        System.out.println("main test: " + (System.currentTimeMillis() - start) / 1000. + "s");
```
Comment (danielezonca) [inline explainability-service/src/test/java/org/kie/trustyai/service/data/storage/performance/PerformanceTest.java:50]: Moreover as far as I see this is just logging the value without any assertion/comparison

File: explainability-service/src/main/resources/application.properties [hunk 9]
```diff
@@ -45,4 +47,25 @@ quarkus.kubernetes-client.devservices.enabled=false
...
 
 # HTTP
-quarkus.http.handle-100-continue-automatically=true
\ No newline at end of file
+quarkus.http.handle-100-continue-automatically=true
+
+# Database
+quarkus.hibernate-orm.active=${HIBERNATE:false}
+quarkus.datasource.db-kind=h2
+quarkus.datasource.jdbc.url=jdbc:h2:file:./trustyai_test_H2_DB
```
Comment (danielezonca) [inline explainability-service/src/main/resources/application.properties:55]: This should be probably moved under the test profile

File: explainability-service/src/main/resources/application.properties [hunk 10]
```diff
@@ -45,4 +47,25 @@ quarkus.kubernetes-client.devservices.enabled=false
...
+#quarkus.hibernate-orm.log.sql=true
+#quarkus.datasource.username=${QUARKUS_DATASOURCE_USER:quarkus}
+#quarkus.datasource.password=${QUARKUS_DATASOURCE_PASSWORD:quarkus}
+#quarkus.datasource.jdbc.url=${QUARKUS_DATASOURCE_URL}
+quarkus.datasource.jdbc.max-size=16
+quarkus.hibernate-orm.database.generation=drop-and-create
```
Comment (danielezonca) [inline explainability-service/src/main/resources/application.properties:61]: Same for this, it is a dangerous option

File: explainability-service/src/main/java/org/kie/trustyai/service/data/storage/hibernate/HibernateStorageInterface.java [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+import org.kie.trustyai.service.data.exceptions.StorageWriteException;
+import org.kie.trustyai.service.data.metadata.StorageMetadata;
+import org.kie.trustyai.service.data.storage.StorageInterface;
+
+import io.quarkus.cache.CacheResult;
+
+public interface HibernateStorageInterface extends StorageInterface {
+    // dataframes
+    @CacheResult(cacheName = "dataframe", keyGenerator = DataCacheKeyGen.class)
+    Dataframe readData(String modelId) throws StorageReadException;
```
Comment (danielezonca) [inline explainability-service/src/main/java/org/kie/trustyai/service/data/storage/hibernate/HibernateStorageInterface.java:15]: This is mixing storage and caching, we need to pay attention to this while we do performance test otherwise we have wrong/fake result

File: explainability-core/src/main/java/org/kie/trustyai/explainability/model/dataframe/Dataframe.java [hunk 12]
```diff
@@ -1161,7 +1207,7 @@ public String toString() {
         for (int i = 0; i < getColumnDimension(); i++) {
 
             builder.append("\tColumn ").append(i);
-            if (metadata.getNameAlias(i) != null) {
+            if (!metadata.getNameAlias(i).equals("")) {
```
Comment (danielezonca) [inline explainability-core/src/main/java/org/kie/trustyai/explainability/model/dataframe/Dataframe.java:1210]:
```suggestion
            if (!"".equals(metadata.getNameAlias(i)) {
```

File: explainability-service/demo/compose-partial-memory-multi-model.yaml [hunk 13]
```diff
@@ -1,6 +1,6 @@
 services:
   trustyai:
-    image: trustyai/trustyai-service:999-SNAPSHOT
+    image: quay.io/rgeada/trustyai-service-debug:latest
```
Comment (danielezonca) [inline explainability-service/demo/compose-partial-memory-multi-model.yaml:3]: To be changed before merging

File: explainability-service/src/main/java/org/kie/trustyai/service/data/storage/Storage.java [hunk 14]
```diff
@@ -1,11 +1,13 @@
 package org.kie.trustyai.service.data.storage;
 
-import java.nio.file.Path;
+import org.kie.trustyai.service.data.exceptions.StorageWriteException;
 
 public abstract class Storage implements StorageInterface {
+    public DataFormat getDataFormat() {
+        return DataFormat.CSV;
```
Comment (danielezonca) [inline explainability-service/src/main/java/org/kie/trustyai/service/data/storage/Storage.java:7]: Is there a reason why we default to `CSV`?

File: explainability-core/src/main/java/org/kie/trustyai/explainability/model/dataframe/DataframeColumn.java (hunk 4 above)
Comment (RobGeada) [reply explainability-core/src/main/java/org/kie/trustyai/explainability/model/dataframe/DataframeColumn.java:10]: removed

File: explainability-core/src/main/java/org/kie/trustyai/explainability/model/dataframe/DataframeInternalData.java (hunk 5 above)
Comment (RobGeada) [reply explainability-core/src/main/java/org/kie/trustyai/explainability/model/dataframe/DataframeInternalData.java:8]: done

File: explainability-core/src/main/java/org/kie/trustyai/explainability/model/dataframe/DataframeMetadata.java (hunk 6 above)
Comment (RobGeada) [reply explainability-core/src/main/java/org/kie/trustyai/explainability/model/dataframe/DataframeMetadata.java:126]: done

File: explainability-service/src/test/java/org/kie/trustyai/service/data/storage/performance/PerformanceTest.java (hunk 7 above)
Comment (RobGeada) [reply explainability-service/src/test/java/org/kie/trustyai/service/data/storage/performance/PerformanceTest.java:15]: removed

File: explainability-service/src/test/java/org/kie/trustyai/service/data/storage/performance/PerformanceTest.java (hunk 8 above)
Comment (RobGeada) [reply explainability-service/src/test/java/org/kie/trustyai/service/data/storage/performance/PerformanceTest.java:50]: done

File: explainability-service/src/main/resources/application.properties (hunk 10 above)
Comment (RobGeada) [reply explainability-service/src/main/resources/application.properties:61]: done

File: explainability-service/src/main/resources/application.properties (hunk 9 above)
Comment (RobGeada) [reply explainability-service/src/main/resources/application.properties:55]: done

File: explainability-service/src/main/java/org/kie/trustyai/service/data/storage/hibernate/HibernateStorageInterface.java (hunk 11 above)
Comment (RobGeada) [reply explainability-service/src/main/java/org/kie/trustyai/service/data/storage/hibernate/HibernateStorageInterface.java:15]: Yep, I'm aware of this, that's why performance tests always retrieve novel dfs, never the same df twice

File: explainability-service/demo/compose-partial-memory-multi-model.yaml (hunk 13 above)
Comment (RobGeada) [reply explainability-service/demo/compose-partial-memory-multi-model.yaml:3]: done

File: explainability-service/src/main/java/org/kie/trustyai/service/data/storage/Storage.java (hunk 14 above)
Comment (RobGeada) [reply explainability-service/src/main/java/org/kie/trustyai/service/data/storage/Storage.java:7]: nope, that's not needed, removed
